pkg/runtime/reconciler.go

```diff
@@ -324,6 +329,40 @@ func (r *resourceReconciler) updateResource(
 	return latest, nil
 }
 
+// lateInitializeResource calls AWSResourceManager.LateInitialize() method and
+// returns the AWSResource with late initialized fields.
+//
+// When the late initialization is delayed for an AWSResource, an error is returned
+// with specific requeue delay to attempt lateInitialization again.
+//
+// This method also adds an annotation to K8s CR, indicating the number of
+// late initialization attempts to correctly calculate exponential backoff delay
+//
+// This method also adds Condition to CR's status indicating status of late initialization.
+func (r *resourceReconciler) lateInitializeResource(
+	ctx context.Context,
+	rm acktypes.AWSResourceManager,
+	desired acktypes.AWSResource,
+	latest acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("r.lateInitializeResource")
+	defer exit(err)
+
+	rlog.Enter("rm.LateInitialize")
+	lateInitializedLatest, err := rm.LateInitialize(ctx, latest)
+	rlog.Exit("rm.LateInitialize", err)
+	if err != nil {
+		// If there was error in late initialization, still patch the resource metadata and spec
+		// to reflect changes in k8s resource.
+		if ackcompare.IsNotNil(lateInitializedLatest) {
+			_ = r.patchResourceMetadataAndSpec(ctx, desired, lateInitializedLatest)
```

Comment:
If `LateInitiialize` set fields on the Spec and did *not* return an error, then we'll never save those changes back to the Kubernetes API server because we'll never call `patchResourceMetadataAndSpec`...

That's why I think this code needs to be:

```go
	rlog.Enter("rm.LateInitialize")
	lateInitializedLatest, err := rm.LateInitialize(ctx, latest)
	rlog.Exit("rm.LateInitialize", err)
	if ackcompare.IsNotNil(lateInitializedLatest) {
		_ = r.patchResourceMetadataAndSpec(ctx, latest, lateInitializedLatest)
	}
	return lateInitializedLatest, err
```

**NOTE**: I'm passing `latest` and `lateInitializedLatest` to `patchResourceMetadataAndSpec`, not `desired` and `lateInitializedLatest`.

Reply:
Excellent point. Making the change now and testing by creating a resource.
IDK how I missed it. Maybe I was relying on the fact that resource was being completely patched at the end of reconciler loop but then we recently did some refactoring and that wasn't the case.
Note:  I will be testing along with latest changes from the main branch.

Reply:
Tested now. I will be updating the PR very soon now.

```
2021-08-11T10:03:55.318-0700    DEBUG   ackrt   >> r.lateInitializeResource     {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}
2021-08-11T10:03:55.318-0700    DEBUG   ackrt   >>> rm.LateInitialize   {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}
2021-08-11T10:03:55.318-0700    DEBUG   ackrt   >>>> rm.sdkFind {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}
2021-08-11T10:03:55.345-0700    DEBUG   ackrt   <<<< rm.sdkFind {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}
2021-08-11T10:03:55.345-0700    DEBUG   ackrt   <<< rm.LateInitialize   {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}
2021-08-11T10:03:55.345-0700    DEBUG   ackrt   >>> r.patchResourceMetadataAndSpec      {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}
2021-08-11T10:03:55.345-0700    DEBUG   ackrt   no difference found between metadata and spec for desired and latest object.    {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}
2021-08-11T10:03:55.345-0700    DEBUG   ackrt   <<< r.patchResourceMetadataAndSpec      {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}
2021-08-11T10:03:55.345-0700    DEBUG   ackrt   << r.lateInitializeResource     {"kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2, "account": "309117047740", "role": "", "region": "us-west-2", "is_adopted": false, "kind": "Repository", "namespace": "default", "name": "some-repo-xyz", "generation": 2}

```

```
spec:
  encryptionConfiguration:
    encryptionType: AES256
  imageScanningConfiguration:
    scanOnPush: false
  imageTagMutability: MUTABLE
  name: some-repo-xyz
status:
  ackResourceMetadata:
    arn: arn:aws:ecr:us-west-2:3091*****:repository/some-repo-xyz
    ownerAccountID: "3091*****"
  conditions:
  - lastTransitionTime: "2021-08-11T17:03:55Z"
    message: Late initialization successful
    reason: Late initialization successful
    status: "True"
    type: ACK.LateInitialized
  createdAt: "2021-08-11T17:03:55Z"
  registryID: "3091*****"
  repositoryURI: 3091*****.dkr.ecr.us-west-2.amazonaws.com/some-repo-xyz
```